### PR TITLE
OAK-10739 : Provide support for FullGC in Mongo DocumentStore

### PR DESCRIPTION
**Pull request by rishabhdaim**

### CODE REVIEW

**Comment by stefan-egli** [comment]
Note that the plan is to do a "merge commit" for this (i.e. _not_ via squash), as it has a lot of useful commit history.

**Comment by stefan-egli** [comment]
Also note that this PR brings in the fullGC functionality completely disabled by default. There will still be more hardening time needed - during which this feature is disabled by default in trunk. It requires to be explicitly be enabled by a feature toggle.

File: oak-jcr/pom.xml [hunk 1]
```diff
@@ -530,5 +530,11 @@
       <artifactId>testcontainers</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (reschke) [inline oak-jcr/pom.xml]: This looks weird; why do we need the duplicate dependency?

File: oak-parent/pom.xml [hunk 2]
```diff
@@ -718,6 +718,13 @@
         <version>${testcontainers.version}</version>
         <scope>test</scope>
       </dependency>
+      <!-- Testcontainers dependency -->
+      <dependency>
+        <groupId>org.testcontainers</groupId>
+        <artifactId>testcontainers</artifactId>
+        <version>${testcontainers.version}</version>
+        <scope>test</scope>
+      </dependency>
```
Comment (reschke) [inline oak-parent/pom.xml]: same here

File: oak-pojosr/pom.xml [hunk 3]
```diff
@@ -365,5 +365,11 @@
       <artifactId>testcontainers</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.testcontainers</groupId>
+      <artifactId>testcontainers</artifactId>
+      <version>${testcontainers.version}</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (reschke) [inline oak-pojosr/pom.xml]: same here

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java [hunk 4]
```diff
@@ -178,7 +183,7 @@ public static NodeStore bootstrapNodeStore(NodeStoreOptions options, Closer clos
             System.exit(1);
         }
 
-        if (src.startsWith(MongoURI.MONGODB_PREFIX) || src.startsWith("jdbc")) {
+        if (src.startsWith(MONGODB_PREFIX) || src.startsWith("jdbc")) {
```
Comment (reschke) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java]: maybe avoid unneeded change here?

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java [hunk 5]
```diff
@@ -210,27 +215,26 @@ static DocumentNodeStoreBuilder<?> createDocumentMKBuilder(String[] args, Closer
         return createDocumentMKBuilder(new NodeStoreOptions(h).parse(args), closer);
     }
 
+    static Optional<MongoConnection> getMongoConnection(final NodeStoreOptions options, final Closer closer) {
+        String src = options.getStoreArg();
+        if (isEmpty(src) || !src.startsWith(MONGODB_PREFIX)) {
```
Comment (reschke) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java]: why do we need isEmpty here? Wouldn't a null check be sufficient?

File: oak-segment-azure/src/test/resources/logback-test.xml [hunk 6]
```diff
@@ -21,11 +21,19 @@
             <pattern>%d{yyyy-MM-dd HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
         </encoder>
     </appender>
-    <statusListener class="ch.qos.logback.core.status.NopStatusListener" />
+    <statusListener class="ch.qos.logback.core.status.NopStatusListener"/>
```
Comment (reschke) [inline oak-segment-azure/src/test/resources/logback-test.xml]: why do we need these changes for a change in DocumentStore?

File: oak-store-document/pom.xml [hunk 7]
```diff
@@ -349,5 +349,11 @@
       <version>${testcontainers.version}</version>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (reschke) [inline oak-store-document/pom.xml]: (indentation)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 8]
```diff
@@ -290,6 +294,25 @@ public boolean isThrottlingEnabled() {
...
+    public T setEmbeddedVerificationEnabled(boolean b) {
+        this.embeddedVerificationEnabled = b;
+        return thisBuilder();
+    }
+
+    public boolean isEmbeddedVerificationEnabled() {
+        return this.embeddedVerificationEnabled;
+    }
+
+
```
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: (vertical whitespace)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java [hunk 9]
```diff
@@ -19,12 +19,28 @@
...
 
+import static java.util.Comparator.comparing;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Stream.concat;
+import static java.util.stream.StreamSupport.stream;
 import static org.apache.jackrabbit.guava.common.collect.Iterables.filter;
+import static java.util.stream.Collectors.toList;
+import static org.apache.jackrabbit.guava.common.collect.Sets.newHashSet;
```
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java]: (avoid use of Guava in new code)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/LoggingDocumentStoreWrapper.java [hunk 10]
```diff
@@ -283,6 +283,7 @@ public T call() throws Exception {
     }
 
     @Override
+    @NotNull
```
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/LoggingDocumentStoreWrapper.java]: is this change needed?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupportTest.java [hunk 11]
```diff
@@ -18,42 +18,54 @@
...
-    private List<String> ids = Lists.newArrayList();
+    private final DocumentStoreFixture fixture;
+    private final DocumentStore store;
+    private final VersionGCSupport gcSupport;
+    private final List<String> ids = new ArrayList<>();
 
     @Parameterized.Parameters(name="{0}")
     public static java.util.Collection<DocumentStoreFixture> fixtures() {
-        List<DocumentStoreFixture> fixtures = Lists.newArrayList();
+        List<DocumentStoreFixture> fixtures = new ArrayList<>(3);
```
Comment (reschke) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupportTest.java]:
```suggestion
        List<DocumentStoreFixture> fixtures = new ArrayList<>();
```

I don't think perf is of any concern here.

**Comment by reschke** [comment]
I started to read, and it's a big change.

Without knowing the details of the design beforehand, it's hard to understand what it does (but that's just a consequence of the complexity of the system).

That said, it would be cool if we could trim down changes to the bare minimum; I see quite a few cosmetic changes (whitespace, static imports, final keywords etc) that do not seem to have anything to do with the change. It would be good if we could reduce the amount of these changes (and yes, that may be in conflict with the goal of not to
 squash the commits)

File: oak-jcr/pom.xml (hunk 1 above)
Comment (stefan-egli) [reply oak-jcr/pom.xml]: agree, looks unnecessary

**Comment by stefan-egli** [comment]
Agree, I think we started letting some minor, unrelated improvements slip in, and we shouldn't have. I'm +1 for removing them.

**Comment by stefan-egli** [comment]
covered the first few comments (more of same kind will still be required though) in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-jcr/pom.xml (hunk 1 above)
Comment (stefan-egli) [reply oak-jcr/pom.xml]: removed in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-parent/pom.xml (hunk 2 above)
Comment (stefan-egli) [reply oak-parent/pom.xml]: removed in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-pojosr/pom.xml (hunk 3 above)
Comment (stefan-egli) [reply oak-pojosr/pom.xml]: removed in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java (hunk 4 above)
Comment (stefan-egli) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/Utils.java]: re-added in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-segment-azure/src/test/resources/logback-test.xml (hunk 6 above)
Comment (stefan-egli) [reply oak-segment-azure/src/test/resources/logback-test.xml]: space re-added in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-store-document/pom.xml (hunk 7 above)
Comment (stefan-egli) [reply oak-store-document/pom.xml]: fixed in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 8 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: removed in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java (hunk 9 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java]: removed guava usage in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupportTest.java (hunk 11 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupportTest.java]: did same change in https://github.com/apache/jackrabbit-oak/pull/1459

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/LoggingDocumentStoreWrapper.java (hunk 10 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/LoggingDocumentStoreWrapper.java]: IntelliJ warns about this missing `@NotNull` annotation since the interface declared this method `@NotNull`.

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/LoggingDocumentStoreWrapper.java]: it seems to be a preexisting issue though, unrelated to fullGC? I'd prefer to remove it.

**Comment by stefan-egli** [comment]
I believe most of the cosmetic changes have been reverted and the PR is ready for review again.

**Comment by mbaedke** [comment]
@stefan-egli, is it really a good idea to do a merge commit? These are 160 commits and I would assume that many of them are intermediate changes or even reverted later. If we do no delete the branch, nothing gets lost, even with a squeeze commit.

**Comment by stefan-egli** [comment]
@mbaedke  yea I had second thought there as well. I'm fine with squeeze as well. @rishabhdaim , @Joscorbe ?

**Comment by rishabhdaim** [comment]
> @mbaedke yea I had second thought there as well. I'm fine with squeeze as well. @rishabhdaim , @Joscorbe ?

I am fine with squashing too.

**Comment by stefan-egli** [comment]
previous, direct commit was just a fix for a previous revert - aka a detail

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 12]
```diff
@@ -324,6 +346,11 @@ public T setNoChildOrderCleanupFeature(@Nullable Feature noChildOrderCleanupFeat
         return thisBuilder();
     }
 
+    public T setDocStoreFullGCFeature(@Nullable Feature docStoreFullGC) {
+        this.docStoreFullGCFeature = docStoreFullGC;
+        return thisBuilder();
+    }
+
```
Comment (mbaedke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: That's out of place. I'd suggest moving it down a few lines to where it belongs.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 13]
```diff
@@ -167,6 +169,8 @@ public class DocumentNodeStoreBuilder<T extends DocumentNodeStoreBuilder<T>> {
     private Predicate<Path> nodeCachePredicate = Predicates.alwaysTrue();
     private boolean clusterInvisible;
     private boolean throttlingEnabled;
+    private boolean fullGCEnabled;
```
Comment (mbaedke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:172]: Not sure if it matters, but why don't we set it to the default value, too?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java [hunk 14]
```diff
@@ -225,6 +236,8 @@ static DocumentStoreType fromString(String type) {
     private Feature docStoreThrottlingFeature;
     private Feature noChildOrderCleanupFeature;
     private Feature cancelInvalidationFeature;
+    private Feature docStoreFullGCFeature;
+    private Feature docStoreEmbeddedVerificationFeature;
```
Comment (mbaedke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:240]: Is this feature ever initialized?

Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:240]: It seems to be a miss, I will initialize this feature.

Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:240]: Addressed in https://github.com/apache/jackrabbit-oak/pull/1482/commits/6b699d245d42d4272e5d7ba4e833d195e7586706

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 12 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: Addressed in https://github.com/apache/jackrabbit-oak/pull/1482/commits/6b699d245d42d4272e5d7ba4e833d195e7586706

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 13 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:172]: We can set them, but the existing pattern is not to set them.

Comment (mbaedke) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:172]: That is true, but then I guess we shouldn't set the default for embeddedVerificationEnabled either.

Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:172]: > That is true, but then I guess we shouldn't set the default for embeddedVerificationEnabled either.

It is set cause the default value for this is `true`; if not initialized, it would be set to `false`.

IIUC, this could be the reason for not setting the default for many booleans (whose default value is `false`) cause that would be assigned by JVM.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java [hunk 15]
```diff
@@ -26,6 +26,8 @@
 import org.apache.jackrabbit.oak.plugins.document.DocumentStore;
 import org.apache.jackrabbit.oak.plugins.document.MissingLastRevSeeker;
 import org.apache.jackrabbit.oak.plugins.document.VersionGCSupport;
+import org.apache.jackrabbit.oak.spi.toggle.Feature;
+import org.jetbrains.annotations.Nullable;
```
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:31]: just checking: why do we need the "Feature" import?

**Review by reschke — APPROVED** [review]
The PR looks much better now. I'm +1 on merging now to avoid having to deal with bit rot again.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 13 above)
Comment (mbaedke) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:172]: Yeah, obviously, that wasn't a particularly bright comment of mine :)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java (hunk 15 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:31]: We need it for overridding `public RDBDocumentNodeStoreBuilder setDocStoreFullGCFeature(@Nullable Feature docStoreFullGC)` method.

**Comment by Joscorbe** [comment]
I'm fine with squashing too @stefan-egli @rishabhdaim.

**Review by Joscorbe — APPROVED** [review]

**Comment by stefan-egli** [comment]
@Joscorbe I see NodeDocumentRevisionCleanerTest.testRecentRevisionsArePreserved failing - would it be much effort to fix that before the merge?

**Review by stefan-egli — APPROVED** [review]
